projects/core/src/lib/classes/mask-model/mask-model.ts

```diff
@@ -1,54 +1,58 @@
-import {MaskExpression} from '../../types';
+import {ElementState, MaskExpression, MaskitoOptions} from '../../types';
 import {validateValueWithMask} from './utils/validate-value-with-mask';
 import {removeFixedMaskCharacters} from './utils/remove-fixed-mask-characters';
 import {addFixedMaskCharacters} from './utils/add-fixed-mask-characters';
 
 export class MaskModel {
     value = '';
+    // TODO: refactor code and replace `caretIndex: number` with `selection: [from: number, to: number]`
     caretIndex = 0;
 
+    private get maskExpression(): MaskExpression {
+        const {mask} = this.maskOptions;
+
+        return typeof mask === 'function' ? mask() : mask;
+    }
+
     constructor(
-        readonly initialValue: string,
-        readonly initialCaretIndex: number,
-        private readonly mask: MaskExpression,
+        readonly initialElementState: ElementState,
+        private readonly maskOptions: MaskitoOptions,
     ) {
-        const {maskedValue, maskedCaretPosition} = addFixedMaskCharacters(
-            initialValue,
-            this.mask,
-            initialCaretIndex,
+        const {value, selection} = addFixedMaskCharacters(
+            initialElementState,
+            this.maskExpression,
         );
 
-        this.value = maskedValue;
-        this.caretIndex = maskedCaretPosition;
+        this.value = value;
+        this.caretIndex = selection[0];
     }
 
-    addCharacters(
-        selectedRange: [from: number, to: number],
-        newCharacters: string,
-    ): void {
-        const {unmaskedValue, unmaskedSelection} = removeFixedMaskCharacters(
-            this.value,
-            this.mask,
-            selectedRange,
+    addCharacters(selection: [from: number, to: number], newCharacters: string): void {
+        const unmaskedElementState = removeFixedMaskCharacters(
+            {value: this.value, selection},
+            this.maskExpression,
```

Comment:
You invoke this getter multiple times and it might be expensive, depending on the mask function. Let's save it to a const.